### ISSUE TITLE
[BUG] ScanJson and JsonToStructs can give almost random errors

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#10633]
**Describe the bug**
When reading some randomly generated JSON data the GPU versions of ScanJson and JsonToStructs can give almost random errors. Some times they work. Other times they fail, and when they do fail the exact error message can change.

**Steps/Code to reproduce bug**
I have an 800 MiB file of randomly generated JSON data. I am happy to share it, but it is kind of large.

```
spark.conf.set("spark.rapids.sql.format.json.enabled", true)
spark.conf.set("spark.rapids.sql.format.json.read.enabled", true)
spark.conf.set("spark.sql.files.maxPartitionBytes", "1g")

spark.read.schema("key_0_0 string").json("/data/tmp/TEST.json").selectExpr("SUM(octet_length(key_0_0))").show()
+--------------------------+
|sum(octet_length(key_0_0))|
+--------------------------+
|                 474605498|
+--------------------------+
spark.read.schema("key_0_0 string").json("/data/tmp/TEST.json").selectExpr("SUM(octet_length(key_0_0))").show()
24/03/25 20:31:16 ERROR Executor: Exception in task 0.0 in stage 16.0 (TID 15)
org.apache.spark.SparkException: Encountered error while reading file file:///data/tmp/TEST.json. Details:
	at org.apache.spark.sql.errors.QueryExecutionErrors$.cannotReadFilesError(QueryExecutionErrors.scala:877)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:307)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:125)
	at org.apache.spark.sql.rapids.GpuFileSourceScanExec$$anon$1.hasNext(GpuFileSourceScanExec.scala:474)
	at com.nvidia.spark.rapids.AbstractProjectSplitIterator.hasNext(basicPhysicalOperators.scala:233)
...
Caused by: ai.rapids.cudf.CudfException: CUDF failure at: ...cudf/cpp/src/io/json/json_column.cu:657: Parent column not found
	at ai.rapids.cudf.Table.readJSON(Native Method)
	at ai.rapids.cudf.Table.readJSON(Table.java:1409)
	at ai.rapids.cudf.Table.readJSON(Table.java:1388)
	at org.apache.spark.sql.catalyst.json.rapids.JsonPartitionReader$.$anonfun$readToTable$2(GpuJsonScan.scala:325)

spark.conf.set("spark.sql.files.maxPartitionBytes", "20m")
24/03/25 20:32:47 ERROR Executor: Exception in task 0.0 in stage 27.0 (TID 26)
org.apache.spark.SparkException: Encountered error while reading file file:///data/tmp/TEST.json. Details:
	at org.apache.spark.sql.errors.QueryExecutionErrors$.cannotReadFilesError(QueryExecutionErrors.scala:877)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:307)
...
Caused by: java.lang.IllegalStateException: Don't know how to transform ColumnVector{rows=548486, type=STRUCT, nullCount=Optional.empty, offHeap=(ID: 178 7fe4000761c0)} to StringType for JSON
  at org.apache.spark.sql.rapids.GpuJsonReadCommon$.throwMismatchException(GpuJsonReadCommon.scala:271)
  at org.apache.spark.sql.rapids.GpuJsonReadCommon$.nestedColumnViewMismatchTransform(GpuJsonReadCommon.scala:304)
  at org.apache.spark.sql.rapids.GpuJsonReadCommon$.$anonfun$convertToDesiredType$1(GpuJsonReadCommon.scala:315)
  at com.nvidia.spark.rapids.ColumnCastUtil$.$anonfun$deepTransformView$2(ColumnCastUtil.scala:79)
  at scala.Option.map(Option.scala:230)
  at com.nvidia.spark.rapids.ColumnCastUtil$.$anonfun$deepTransformView$1(ColumnCastUtil.scala:77)
  at com.nvidia.spark.rapids.Arm$.closeOnExcept(Arm.scala:126)
  at com.nvidia.spark.rapids.ColumnCastUtil$.deepTransformView(ColumnCastUtil.scala:62)
  at com.nvidia.spark.rapids.ColumnCastUtil$.deepTransform(ColumnCastUtil.scala:183)
  at org.apache.spark.sql.rapids.GpuJsonReadCommon$.convertToDesiredType(GpuJsonReadCommon.scala:315)
  at org.apache.spark.sql.rapids.GpuJsonReadCommon$.$anonfun$convertTableToDesiredType$2(GpuJsonReadCommon.scala:361)
  at com.nvidia.spark.rapids.RapidsPluginImplicits$MapsSafely.$anonfun$safeMap$1(implicits.scala:221)
  at com.nvidia.spark.rapids.RapidsPluginImplicits$MapsSafely.$anonfun$safeMap$1$adapted(implicits.scala:218)

```

**Expected behavior**
If it works, it should always work. I have been trying to come up with a CUDF only repro case, but I have not been able to do so yet.

**Comment by revans2** [comment]
The could not find parent error appears to be related to the order that some columns show up into the code in question.

https://github.com/rapidsai/cudf/blob/933e32ab9ad8e5057282c48129ddbd745c538967/cpp/src/io/json/json_column.cu#L657

Appears to be related to the order that `unique_col_ids` is passed into the function. If the parent column is not before the children this will happen. I don't know why the order might be different, but this appears to happen regularly.

**Comment by revans2** [comment]
Looks like you need to run it more than once.

```
import ai.rapids.cudf._

val sb = Schema.builder()

sb.addColumn(DType.STRING, "key_0_0")

val s = sb.build

val opts = JSONOptions.builder.withKeepQuotes(true).withLines(true).withNormalizeSingleQuotes(true).withRecoverWithNull(true).withNormalizeWhitespace(true).build()

val t = Table.readJSON(s, opts, new java.io.File("/data/tmp/TEST.json"))

t.getRowCount

t.close()

val t = Table.readJSON(s, opts, new java.io.File("/data/tmp/TEST.json"))

ai.rapids.cudf.CudfException: CUDF failure at: .../cudf/cpp/src/io/json/json_column.cu:657: Parent column not found
  at ai.rapids.cudf.Table.readJSON(Native Method)
  at ai.rapids.cudf.Table.readJSON(Table.java:1266)
  ... 49 elided
```

Will try to get a C++ repro case

**Comment by revans2** [comment]
It looks like it is related to memory pooling, and possibly reading uninitialized memory. It works fine if pooling is disabled, but fails regularly if it is enabled. (at least on the java side of thing)

**Comment by revans2** [comment]
I was able to cut the file down to 861262 lines (about 26 MiB) and I am still able to see errors. Will keep working on this...

**Comment by revans2** [comment]
Digging deeper the tokenization is returning a different set of tokens. I am not sure why yet.  The data looks fine for the first part of the run, and then it runs into a point where it looks like there was an error and after that everything is off.

**Comment by revans2** [comment]
A little more info. This is only happening in java on the async allocator. Not the arena. This is all really confusing to me.

**Comment by revans2** [comment]
If I set the config for recover with null to false that appears to fix the problem. Recover with nulls is odd because it is updating the data inline in the device buffer. I am not 100% sure what it is doing, or how it deals with the end of the record.

**Comment by revans2** [comment]
@GregoryKimball @shrshi  I really would appreciate some help in understanding what the next steps should be for debugging this.  I have a test case that I can repro nearly 100% of the time on my desktop in java/scala, but I cannot do it in C++. The error appears to be related to the Async Allocator in RMM and when setting recover_with_nulls to true (For java we install some other layers to the async allocator so it could be changing the timing for an allocation/free).

I don't understand the internals of the JSON parser much at all, but I am concerned that the `Transduce` call for recover from errors has an input `stack_symbols`, that is also an output. I don't know if `Transduce` is really setup to do in-place modification or not.  But I also don't know how the Async allocator could be causing the issue, unless it is just around timing and there is a race that it makes us lose consistently, or there is a serious bug in the async allocator that only shows up when run in a JVM.

**Comment by revans2** [comment]
Okay it is a race somewhere. I put in a bunch of `stream.synchronize` calls in the JSON parsing code and the problem appears to have gone away. I will try and narrow it down more to get a better idea where the problems might be.

**Comment by revans2** [comment]
For some reason the following fixes the issue.

```
diff --git a/cpp/src/io/json/nested_json_gpu.cu b/cpp/src/io/json/nested_json_gpu.cu
index a6a57c36b0..834f5d73de 100644
--- a/cpp/src/io/json/nested_json_gpu.cu
+++ b/cpp/src/io/json/nested_json_gpu.cu
@@ -1583,6 +1583,8 @@ std::pair<rmm::device_uvector<PdaTokenT>, rmm::device_uvector<SymbolOffsetT>> ge
                                         thrust::make_discard_iterator(),
                                         fix_stack_of_excess_chars::start_state,
                                         stream);
+    // Failure happens without this sync? Not sure why
+    stream.synchronize();
   }
 
   constexpr auto max_translation_table_size =
```

I want to understand why it fixes the issue, but I don't want to lose this.

**Comment by abellina** [comment]
@revans2 asked us to take a look since he's going to be out for next week and I was able to reproduce the failure he is seeing. @jlowe and I look at this and the pinned memory pool changed the behavior (no pinned memory = it works, pinned memory = it doesn't work).

The specific case @revans2 pointed at here https://github.com/NVIDIA/spark-rapids/issues/10633#issuecomment-2023134292 is a problem because with the pinned pool we no longer have a `cudaFreeHost` synchronization, since the host memory is now pooled. In other words, we think the old code was relying on implicit synchronization.

The synchronize Bobby added looks to help because the destructor for `Dfa` has to wait for it on the stream we used to launch kernels from the `Dfa::Transduce` call. The destructor is getting triggered after we exit this scope https://github.com/rapidsai/cudf/blob/branch-24.06/cpp/src/io/json/nested_json_gpu.cu#L1572-L1586 for the object `fix_stack_of_excess_chars`. 

This object is of type `Dfa` and it contains a `hostdevice_vector` https://github.com/rapidsai/cudf/blob/branch-24.06/cpp/src/io/fst/lookup_tables.cuh#L899. This `hostdevice_vector` is now using the new cuIO pinned pool work, so a free returns the object to the pool and doesn't incur extra synchronization.
